File: 0_string/reorganize_str_767.py
'''
input: string s
goal: rearrange the characters of s so that any two adjacent characters are not the same 
output: rearranged string 

ex) 
s = "aab" 
output = "aba" 

s = "aaab" 
output = "" 

'''


import heapq
import collections 

class Solution:
    def reorganizeString(self, s: str) -> str:
        if not s: return ""

        #create a counter 
        heap = []
        # store the letter and frequence as pair based on its frequence in max heap
        # key = the letter and value = frequence 
        # counter = dictionary form 
        # https://www.daleseo.com/python-heapq/ 
        # min heap이 기본이라, max heap을 해주려면 tuple을 원소로 추가하거나 삭제하면, 튜플내에서 맨 앞에 있는 값을 기준으로 min heap이 구성되는 원리를 이용한다. 
        # 인덱스 0은 우선순위이므로 나중에 값을 가져올 때는 index 1에 있는 값만 가져오면 된다. 
        for key, value in collections.Counter(s).items():
           heapq.heappush(heap, [-value, key])
        
        # res will be pre + curr 
        # if s = "aab", res = 'a' + 'b' + 
        res = ""
        pre = heapq.heappop(heap)
        res += pre[1]

        while heap:
            curr = heapq.heappop(heap)
            res += curr[1] 

            pre[0] += 1 # max heap을 이용했으므로 +1 를 해줘야 0에 가까워짐 
            
            # 아직 횟수가 남았으므로 heap에 다시 넣어줌 
            if pre[0] <0:
                heapq.heappush(heap, pre)
            
            pre = curr 

        return "" if len(res) != len(s) else res 
    

if __name__ == '__main__':
    s = "aaabbcd"
    x = Solution()
    print(x.reorganizeString(s)) # babacd 
    

File: 0_string/reorganize_str_767.cpp
/*

input: string s
goal: rearrange the characters of s so that any two adjacent characters are not the same 
output: rearranged string 

ex) 
s = "aab" 
output = "aba" 

s = "aaab" 
output = "" 

*/
#include <iostream>
#include <unordered_map> 
#include <queue> // priority queue , push, pop & top, empty, size 
#include <utility> //pair 
#include <string> // string 
using namespace std;

class Solution {
public:
    string reorganizeString(string s) {
        if (s.empty()) return "";

        // Counter 
        unordered_map<char, int> cnt;
        
        for (char c : s){
            cnt[c]++;
        }

        // max-heap comparator  
        //어떠한 variable에 간단한 function을 assign할 때 주로 사용됨 
        auto cmp = [](pair<char, int> &a, pair<char, int> & b){
            return a.second < b.second; // dictionary의 key를 비교 
        }; // lambda fucntion 뒤에는 ; 를 써야한다!!!

        //decltype : declare type 
        std:priority_queue<pair<char, int>, vector<pair<char, int>>, decltype(cmp)> maxHeap(cmp);
        
        for (auto & entry: cnt){
            maxHeap.push({entry.first, entry.second}); // pair는 make_pair(1, 'a') 아니면 {1, 'a'}로 initialization할 수 있다. 
        }

        // make result 
        string res("");
        pair<char, int> pre = maxHeap.top();
        maxHeap.pop();
        res += pre.first;
       

        while (!maxHeap.empty()){
            pair<char, int> cur = maxHeap.top();
            maxHeap.pop();
            res += cur.first;

            pre.second -= 1;
            // If frequency is not exhausted, push back to the heap 
            if (pre.second > 0){
                maxHeap.push(pre);
            }

            pre = cur; 
        }

        // (condition) ? value_a : value_b; 
        // if the condition is true, return value_a otherwise return value_b
        return (res.length() != s.length()) ? "" : res;



    }
};


int main() {
    Solution sol;
    std::string input = "aab";
    std::string result = sol.reorganizeString(input);
    std::cout << result << std::endl;
    return 0;
}